Extend hierarchy functions to work with specified Solr backend

samuli opened the pull request:
This PR extends hierarchy functions (hierarchy tree and search for Collection records) to work also with other Solr backends than the default one. Currently, search functions in `HierarchyController` always use the default Solr backend configured in `config.ini > Index`. For `Search2` records this means that the index configured in `Search2.ini > Index` is not used in hiearchy searches. This affects also other RecordDrivers and Search/CollectionControllers that extend Solr-classes. 

`HierarchySearch2.ini` should reflect `HierarchyDefault.ini` apart from `treeSource`. How should this be handled in a backward compatible way? Should we include `HierarchySearch2.ini` as a .sample file so that maintainers need to copy it manually and set other settings according to their current `HierarchyDefault.ini`?

Changes to `collectionlist.phtml` and `HierarchyAwareTrait.php` fix undefined variables that I noticed while working on the PR.

demiankatz commented:
@samuli, thanks for this! I probably won't have time for a close review until next week, but in the meantime, to answer your question about HierarchySearch2.ini, did you consider using the [Parent_Config] section to inherit from HierarchyDefault.ini? This would allow you to override only the setting that needs to change while sharing most of the other configuration with the defaults. I think it's also okay to include the fully replicated example, but if you want to emphasize the distinction between the files, the inheritance approach might be a good way to do that.

samuli commented:
@demiankatz Sorry, I remembered that `[Parent_Config]` was implemented only in our extended codebase... Edited to inherit from `HierarchyDefault.ini`

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/Search2Default.php:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @SuppressWarnings(PHPMD.ExcessivePublicCount)
+ */
+class Search2Default extends SolrDefault
```
I think this class should be registered in VuFind\RecordDriver\PluginManager -- it may be working because of default behavior, but it would be best to be more explicit.

samuli replied:
Yes, updated. Maybe it works because `RecordCollectionFactory` uses `getSearch2Record`.

demiankatz commented on module/VuFind/src/VuFind/Hierarchy/TreeRenderer/JSTree.php:
```diff
@@ -249,7 +249,14 @@ protected function getUrlFromRouteCache($route, $id)
                 ]
             ];
-            $cache[$route] = $this->router->fromRoute($route, $params, $options);
+            $routeName = $route;
```
I wonder if this new logic would actually fit better as a support method, both for readability and overriding... e.g.

<pre>
protected function getRouteNameFromDatasource($route)
{
    if ($route === 'collection') {
        return $this->getDataSource->getCollectionRoute();
    }
    if ($route === 'record') {
        return $this->getDataSource->getRecordRoute();
    }
    return $route;
}
</pre>

That's extremely nitpicky, and you can feel free to leave this as-is if you prefer... but that feels marginally better to me.

samuli replied:
I agree, more clear this way. It also provides even more flexibility for extending.

demiankatz commented on module/VuFind/src/VuFind/Controller/Search2recordController.php:
```diff
@@ -65,4 +65,14 @@ protected function resultScrollerActive()
+     *
+     * @return string|bool
+     */
+    public function getHierarchyType()
```
Does this method do anything? I'm not seeing any getHierarchyType method used/defined in the core code's controllers -- this seems to be a record driver mechanism.

samuli replied:
You're right, it's not needed here.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/Search2Default.php:
```diff
@@ -0,0 +1,65 @@
+     */
+    public function getHierarchyType()
+    {
+        return 'search2';
```
I'm not sure what the best/most universal way to reconcile this with [HierarchyAwareTrait::getHierarchyType](https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/RecordDriver/HierarchyAwareTrait.php#L225), since different approaches may be more appropriate for different situations... but all nuances aside, should this at least return false when the hierarchy fields are unset, since presumably many search2 records will exist without hierarchy settings?

samuli replied:
Yes, this should definitely at least check the hierarchy fields. Edited to use `parent::getHierarchyType` and modify the output only when needed.

demiankatz replied:
Thanks!

samuli commented:
@demiankatz I noticed that `Search2Collection` screen was still targeting the normal index. This was due to missing `Search2Collection` specific Options/Params/Results. I added these. Also a specific version of `CollectionList` recordtab is needed. This is now configured to be used with `Search2Default` records in `RecordTabs.ini`.

demiankatz commented on module/VuFind/src/VuFind/Search/Factory/SolrDefaultBackendFactory.php:
```diff
@@ -43,6 +43,13 @@
+     *
+     * @var string
+     */
+    protected $createRecordCallback = 'getSolrRecord';
```
It might be better to call this `$createRecordMethod` since it's not a full PHP callback, just a method name for the plugin manager.

samuli replied:
I agree, renamed.

demiankatz replied:
Thanks!

demiankatz commented on themes/bootstrap3/templates/RecordTab/search2collectionlist.phtml:
```diff
@@ -0,0 +1 @@
+collectionlist.phtml
```
Is there supposed to be more content here? Unless this is using a syntax I'm not familiar with, it seems incomplete.

samuli replied:
It's a symlink between `collectionlist.phtml` > `search2collectionlist.phtml` to avoid duplicating the template. Maybe `RecordTab\Search2CollectionList` could be somehow set to render `collectionlist.phtml` instead.

demiankatz replied:
Try `<?=$this->render('RecordTab/collectionlist.phtml')?>` instead -- that should have the same effect, but in a more platform-independent way.

samuli replied:
Yes, better this way.

demiankatz commented on module/VuFind/src/VuFind/Hierarchy/TreeDataSource/Search2.php:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @var string
+     */
+    protected $collectionRoute = 'search2collection';
```
I notice that you do reference a search2collection route here, but it doesn't seem to be defined elsewhere...

demiankatz commented on module/VuFind/src/VuFind/RecordTab/CollectionList.php:
```diff
@@ -63,6 +63,13 @@ class CollectionList extends AbstractBase
+     *
+     * @var string
+     */
+    protected $backendId = 'SolrCollection';
```
I would actually change this to `$searchClassId` -- SolrCollection is not a backend ID; it's a set of search objects that add extra abstraction on top of the Solr backend. I think using the backend terminology here might add confusion.

samuli replied:
Yes, renamed.

samuli commented:
@demiankatz Yes, I was testing without `collections` and `route[search2record] = search2collection`  and got the same problems as soon as these were set....

1) I added `Search2collectionController` and also had to move collection-link building to `Record` view helper that now renders a collection specific `link-` template. If this is too complicated, another option would be to keep using `url` helper and return the name of the collection route from the record driver. 
All links should now stay within the selected backend, but more testing wouldn't hurt.
While going through the templates using `url` helper to return `collection` route, I noticed the `Collections` controller (for browsing collections) that is not updated to handle Search2 collections... Do you think this should be addressed? I couldn't get any search results when using it with Solr, but it could be some issue related to my environment.

2) I changed Search2 hierarchy cache files to use a prefix. Solr cache files are named as before so existing caches will still work. I also extended `util createHierarchyTrees` to work with `Search2` backend. 
Btw: cache files generated with `createHierarchyTrees` get saved to `/data/local/hierarchy`. Is this on purpose or should they be saved to `/local/cache/hierarchy` (which is used when triggered from the UI).

demiankatz commented:
@samuli, thanks for the continued work on this -- it's one of those projects that keeps getting bigger as you go (though I think we're nearly at the end now). :-)

Regarding the CollectionsController, we would probably need to refactor it to make some things overrideable, then create a Collections2Controller to support all the separate configurations, etc., for the secondary index. That might not be worth the effort at this stage if you don't actually need it for your purposes, but if you want to cover every situation, it could be done.

Using the Record helper for linking collections makes sense to me, but I'll take a closer look at the revised code soon to see if I have any further suggestions.

Regarding the hierarchy files, the command line utility is meant to write to the same cache that gets read by the web interface. Is it possible you're not exporting all of your environment variables when running the utility, and that is causing it to load a different configuration or behave differently? If you're not easily able to figure out what's going on, let me know and I'll see if I'm encountering the same behavior on my end.

demiankatz commented:
You might want to `export VUFIND_HOME=/path/to/vufind ; export VUFIND_LOCAL_DIR=/path/to/vufind/local` ; export VUFIND_LOCAL_MODULES=CustomModuleName` and see if that makes a difference. I'll also try it over here when time permits!

samuli commented:
@demiankatz Ok, with `VUFIND_HOME` and `VUFIND_LOCAL_DIR` set the cache files are saved to `/local/cache/hierarchy`.

demiankatz commented:
Great, thanks, @samuli; sounds like victory is at hand. I'm probably not going to have a chance to do further testing until next week, but I'll see if I can get it merged then!

demiankatz commented:
@samuli, I found time to work on this earlier than expected, and I've merged a few more stand-alone parts of this PR to master (some with minor modifications). At this point, I only have one question: is it strictly necessary to add the $extras parameter to the Record view helper's getLink() method? Couldn't you just pass the collection ID as the $lookfor parameter and work with it that way? I don't think that's too far of a stretch, but maybe I'm overlooking another detail.

samuli commented:
> is it strictly necessary to add the $extras parameter to the Record view helper's getLink() method?

No, not really. Simplified to use collection ID as `$lookfor`.

demiankatz commented:
Thanks, @samuli, this is working well for me now, so I have merged it.

Reviewing and testing this did raise a couple of pre-existing issues that are not impacted by merging this -- I'm wondering if you are seeing the same things, and whether @crhallberg or @EreMaijala might be able to weigh in:

1.) When looking at the "Context" tab for a record in a hierarchy, the folder/file icon disappears next to the active record. I'm guessing there is supposed to be some styling in place to indicate a highlighted row, but I'm not seeing it. I wonder if some CSS got lost or broken somewhere along the way.

2.) I find the breadcrumb template logic extremely confusing. I'm not sure why one template references record routes and the other references collection routes, or under what circumstances these templates get rendered. If memory serves, this was refactored long ago as part of the VuDL effort, which has since been removed; I wonder if this can be simplified in release 7 to prevent future confusion.

Anyway, we should probably take this conversation elsewhere if these items need further discussion, but I wanted to raise them while I was thinking of them. I can open JIRA tickets if anyone thinks it's worth the effort!

EreMaijala commented:
@demiankatz 1.) I don't think the folder or file icons should be visible at all. As far as I can remember the default styles used to include a colored background box around the text to indicate the selected item. In fact we still display it in Finna, see e.g. https://www.finna.fi/Record/narc.VAKKA-310440.KA_VAKKA-239245.KA/HierarchyTree?recordID=narc.VAKKA-310440.KA_VAKKA-239245.KA&lng=en-gb

demiankatz commented:
@EreMaijala, a bit of digging reveals that you removed the styles that put the correct color/background color on the selected hierarchy item in commit 2e6a78ade3cd3dc7e917a6d39ce5166a3c75dab5 -- any idea why? I think we need to put them back in order to make the collection functionality display correctly, but I'm concerned that you may have removed them for a valid reason related to hierarchical facets. Any suggestions?

EreMaijala commented:
@demiankatz Oh no, I was afraid it might have been me. I think we need to restore the style but with a more targeted selector.

demiankatz commented:
Makes sense. @crhallberg, can you look into this? Hopefully it's a quick fix once we decide on the selector.
